### PR TITLE
feat: set legacy decorators version when `experimentalDecorators` enabled

### PR DESCRIPTION
**Pull request by Timeless0911**
## Summary

- use tsconfck to find and parse tsconfig file
- set legacy decorators version when `experimentalDecorators` enabled
- add `tsconfigPath` test cases
- prebundle `rslog`

## Checklist

- [x] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

**Review by fi3ework — COMMENTED** [review]
I just recalled https://github.com/web-infra-dev/modern.js/issues/5834.

How does esbuild do? See https://github.com/evanw/esbuild/blob/main/CHANGELOG.md#0210:

> Note that this feature is different than the existing "TypeScript experimental decorators" feature that esbuild already implements. It uses similar syntax but behaves very differently, and the two are not compatible (although it's sometimes possible to write decorators that work with both). TypeScript experimental decorators will still be supported by esbuild going forward as they have been around for a long time, are very widely used, and let you do certain things that are not possible with JavaScript decorators (such as decorating function parameters). By default esbuild will parse and transform JavaScript decorators, but you can tell esbuild to parse and transform TypeScript experimental decorators instead by setting "experimentalDecorators": true in your tsconfig.json file.

For projects that have `experimentalDecorator` enabled in tsconfig, should we automatically switch to `legacyDecorator` (and `emitDecoratorMetadata`)? Additionally, when we perform checks between configurations later on, `legacyDecorator` should also check whether it is compatible with `decorators.version`.

In create-rslib, I think we're good to use the new version of decorator by default.

**Comment by Timeless0911** [comment]
https://github.com/web-infra-dev/rsbuild/blob/01a8b34143b4957225186e9b80b050103582e546/packages/core/src/plugins/swc.ts#L250-L273

The Rsbuild logic is above, and we should just do a tsconfig detect, and if users config `experimentalDecorator: true`, we automatically inject legacy decorator Rsbuild config.

**Comment by fi3ework** [comment]
> [web-infra-dev/rsbuild@`01a8b34`/packages/core/src/plugins/swc.ts#L250-L273](https://github.com/web-infra-dev/rsbuild/blob/01a8b34143b4957225186e9b80b050103582e546/packages/core/src/plugins/swc.ts#L250-L273)
> 
> The Rsbuild logic is above, and we should just do a tsconfig detect, and if users config `experimentalDecorator: true`, we automatically inject legacy decorator Rsbuild config.

Should this be handled by Rsbuild? I think it also applies to app building.

**Comment by Timeless0911** [comment]
> > [web-infra-dev/rsbuild@`01a8b34`/packages/core/src/plugins/swc.ts#L250-L273](https://github.com/web-infra-dev/rsbuild/blob/01a8b34143b4957225186e9b80b050103582e546/packages/core/src/plugins/swc.ts#L250-L273)
> > The Rsbuild logic is above, and we should just do a tsconfig detect, and if users config `experimentalDecorator: true`, we automatically inject legacy decorator Rsbuild config.
> 
> Should this be handled by Rsbuild? I think it also applies to app building.

Rsbuild does not want to be associated with tsconfig.json fields.

File: packages/core/src/config.ts [hunk 1]
```diff
@@ -10,6 +10,7 @@ import {
   rspack,
 } from '@rsbuild/core';
 import glob from 'fast-glob';
+import ts from 'typescript';
```
Comment (chenjiahan) [inline packages/core/src/config.ts:13]: this will break JS projects

File: packages/core/src/config.ts [hunk 2]
```diff
@@ -360,6 +362,25 @@ export function composeBannerFooterConfig(
...
+  >['version'],
+): RsbuildConfig {
+  if (version || !options.experimentalDecorators) {
+    return {};
+  }
+
+  return {
+    source: {
+      decorators: {
+        version: 'legacy',
```
Comment (chenjiahan) [inline packages/core/src/config.ts:378]: why defaults to legacy

File: packages/core/src/config.ts (hunk 1 above)
Comment (fi3ework) [reply packages/core/src/config.ts:13]: We could use tsconfck to look up tsconfig, which is also used by Vite.

File: packages/core/src/config.ts (hunk 2 above)
Comment (Timeless0911) [reply packages/core/src/config.ts:378]: This is the return value of `1&&2`
1. no `source.decorators.version` config
2. `experimentalDecorators`: true in tsconfig

Comment (Timeless0911) [reply packages/core/src/config.ts:378]: Maybe I should write on the contrary

```ts
  if (!version && options.experimentalDecorators) {
    return {
    source: {
      decorators: {
        version: 'legacy',
      },
    },
  };
  }

  return {};
```

Comment (chenjiahan) [reply packages/core/src/config.ts:378]: My mistake, the current implementation is ok

File: packages/core/src/config.ts (hunk 1 above)
Comment (Timeless0911) [reply packages/core/src/config.ts]: I use `tsconfck` instead,  more details can be implemented when dealing no config feature.

File: packages/core/package.json [hunk 3]
```diff
@@ -51,6 +51,7 @@
     "prebundle": "1.2.2",
     "rslib": "npm:@rslib/core@0.0.5",
     "rslog": "^1.2.3",
+    "tsconfck": "3.1.3",
```
Comment (fi3ework) [inline packages/core/package.json:54]: Do we need to precompile it or using the `autoExternal` feature instead? I don't know how to choose between these two now TBH 😂.

Comment (fi3ework) [reply packages/core/package.json:54]: Got it, tsconfck is an aggressive ESM package without main. I think prebundle should also support `import.meta.resolve()` to resolve ESM entry as well in the future. I also found that rslog is not prebundled, it that on purpose? 🤔

Comment (Timeless0911) [reply packages/core/package.json:54]: Forgot to add rslog before.

**Review by fi3ework — APPROVED** [review]

**Pull request merged**